channels/rdpgfx/client/rdpgfx_main.c

```diff
@@ -689,19 +904,22 @@ static UINT rdpgfx_recv_cache_import_reply_pdu(RDPGFX_CHANNEL_CALLBACK* callback
 	if (!Stream_CheckAndLogRequiredLength(TAG, s, 2ull * pdu.importedEntriesCount))
 		return ERROR_INVALID_DATA;
 
-	pdu.cacheSlots = (UINT16*)calloc(pdu.importedEntriesCount, sizeof(UINT16));
+	if (pdu.importedEntriesCount > RDPGFX_CACHE_ENTRY_MAX_COUNT)
+		return ERROR_INVALID_DATA;
 
-	if (!pdu.cacheSlots)
+	if (Stream_GetRemainingLength(s) / 2 < (size_t)pdu.importedEntriesCount)
 	{
-		WLog_Print(gfx->log, WLOG_ERROR, "calloc failed!");
-		return CHANNEL_RC_NO_MEMORY;
+		WLog_Print(gfx->log, WLOG_ERROR, "not enough data!");
+		return ERROR_INVALID_DATA;
 	}
 
-	for (index = 0; index < pdu.importedEntriesCount; index++)
+	for (idx = 0; idx < (int) pdu.importedEntriesCount; idx++)
 	{
-		Stream_Read_UINT16(s, pdu.cacheSlots[index]); /* cacheSlot (2 bytes) */
+		Stream_Read_UINT16(s, pdu.cacheSlots[idx]); /* cacheSlot (2 bytes) */
 	}
 
+	rdpgfx_load_cache_import_reply(gfx, &pdu);
+
```

Comment:
The result shall be checked, I'm wondering what we must do if the import fails (disconnect or continue ?)

Reply:
I added the proper error check, I'm not sure if we should hard fail on this one